File: android/app/src/main/java/com/synapse/messenger/network/protocol/Caps.kt
package com.synapse.messenger.network.protocol

/**
 * Capability bitset negotiated in HELLO/WELCOME — mirrors
 * `server/pkg/wire/constants.go`. The agreed set is the intersection; unknown
 * bits are ignored on both sides, so the bitset may grow. Never renumber a bit.
 */
object Cap {
    /** peer understands gzip-compressed frames */
    const val COMPRESSION = 1 shl 0

    /** peer understands batched envelopes (reserved on the server; unused) */
    const val BATCHING = 1 shl 1

    /** peer supports session resume */
    const val RESUME = 1 shl 2

    /** peer supports E2E secret chats (X3DH + Double Ratchet) */
    const val SECRET_CHAT = 1 shl 3

    /** peer wants typing/presence events */
    const val TYPING_SIGNALS = 1 shl 4

    /** peer understands zstd frames compressed against the server's shared dictionary */
    const val ZSTD = 1 shl 5

    /**
     * What this client advertises.
     *
     * COMPRESSION is advertised: the gateway gzips outbound frames over 256 bytes
     * for peers that negotiated it, which is worth real bytes on a mobile link,
     * and [Frame] gunzips inbound frames. ZSTD is not — the server compresses
     * against a shared raw dictionary we have no copy of, so a zstd frame would
     * be undecodable here.
     *
     * SECRET_CHAT is omitted because this client implements no Double Ratchet;
     * claiming it would invite ciphertext we cannot decrypt.
     */
    const val CLIENT_CAPS = COMPRESSION or RESUME or TYPING_SIGNALS

    fun has(caps: Int, cap: Int): Boolean = (caps and cap) != 0
}

File: client/src/shared/api/protocol/caps.ts
/**
 * Capability bitset negotiated in HELLO/WELCOME — mirrors
 * `server/pkg/wire/constants.go`.
 *
 * The agreed set is the intersection of what we advertise and what the gateway
 * supports; unknown bits are ignored on both sides, so the bitset can grow
 * without breaking either peer. Never renumber a bit — only append.
 */
export const Cap = {
  /** peer understands gzip-compressed frames */
  COMPRESSION: 1 << 0,
  /** peer understands batched envelopes (reserved; not used yet) */
  BATCHING: 1 << 1,
  /** peer supports session resume */
  RESUME: 1 << 2,
  /** peer supports E2E secret chats */
  SECRET_CHAT: 1 << 3,
  /** peer wants typing/presence events */
  TYPING_SIGNALS: 1 << 4,
  /** peer understands zstd+dictionary frames */
  ZSTD: 1 << 5,
} as const

export type Cap = (typeof Cap)[keyof typeof Cap]

/**
 * What this web client advertises.
 *
 * Compression is deliberately NOT advertised. The gateway compresses an
 * outbound frame only when the peer negotiated a compression capability, so
 * omitting both bits guarantees every frame we receive is raw — which matters
 * because the browser has no zstd decoder (`DecompressionStream` supports
 * gzip/deflate only, and the server's zstd path uses a shared dictionary we
 * could not supply anyway). Advertising CapCompression alone would work via
 * `DecompressionStream('gzip')`, but it buys little for chat-sized frames and
 * costs an async hop on every inbound frame.
 *
 * SECRET_CHAT is omitted because this client implements no Double Ratchet;
 * claiming the capability would invite ciphertext we cannot decrypt.
 */
export const CLIENT_CAPS = Cap.RESUME | Cap.TYPING_SIGNALS

export function hasCap(caps: number, cap: Cap): boolean {
  return (caps & cap) !== 0
}
